# 00）数据导入 读取

（导入基本数据分析模块）（读取原始数据文件内容）

In [1]:
import pandas as pd

In [3]:
original_titles =pd.read_csv("titles.csv")

In [4]:
original_credits = pd.read_csv("credits.csv")

In [5]:
original_titles.head()

id                                title   type  \
0  ts300399  Five Came Back: The Reference Films   SHOW   
1   tm84618                          Taxi Driver  MOVIE   
2  tm154986                          Deliverance  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
4  tm120801                      The Dirty Dozen  MOVIE   

                                         description  release_year  \
0  This collection includes 12 World War II-era p...          1945   
1  A mentally unstable Vietnam War veteran works ...          1976   
2  Intent on seeing the Cahulawassee River before...          1972   
3  King Arthur, accompanied by his squire, recrui...          1975   
4  12 American military prisoners in World War II...          1967   

  age_certification  runtime                                       genres  \
0             TV-MA       51                            ['documentation']   
1                 R      114                           ['drama', 'crime']   
2                 R      109  ['drama', 'action', 'thriller', 'european']   
3                PG       91              ['fantasy', 'action', 'comedy']   
4               NaN      150                            ['war', 'action']   

  production_countries  seasons    imdb_id  imdb_score  imdb_votes  \
0               ['US']      1.0        NaN         NaN         NaN   
1               ['US']      NaN  tt0075314         8.2    808582.0   
2               ['US']      NaN  tt0068473         7.7    107673.0   
3               ['GB']      NaN  tt0071853         8.2    534486.0   
4         ['GB', 'US']      NaN  tt0061578         7.7     72662.0   

   tmdb_popularity  tmdb_score  
0            0.600         NaN  
1           40.965       8.179  
2           10.010       7.300  
3           15.461       7.811  
4           20.398       7.600

In [6]:
original_credits.head()

person_id       id             name                character   role
0       3748  tm84618   Robert De Niro            Travis Bickle  ACTOR
1      14658  tm84618     Jodie Foster            Iris Steensma  ACTOR
2       7064  tm84618    Albert Brooks                      Tom  ACTOR
3       3739  tm84618    Harvey Keitel  Matthew 'Sport' Higgins  ACTOR
4      48933  tm84618  Cybill Shepherd                    Betsy  ACTOR

# 01）评估和清理数据

In [11]:
cleaned_titles=original_titles.copy()

In [12]:
cleaned_credits=original_credits.copy()

## 数据整齐度

In [13]:
cleaned_titles.head(10)

id                                title   type  \
0  ts300399  Five Came Back: The Reference Films   SHOW   
1   tm84618                          Taxi Driver  MOVIE   
2  tm154986                          Deliverance  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
4  tm120801                      The Dirty Dozen  MOVIE   
5   ts22164         Monty Python's Flying Circus   SHOW   
6   tm70993                        Life of Brian  MOVIE   
7   tm14873                          Dirty Harry  MOVIE   
8  tm119281                     Bonnie and Clyde  MOVIE   
9   tm98978                      The Blue Lagoon  MOVIE   

                                         description  release_year  \
0  This collection includes 12 World War II-era p...          1945   
1  A mentally unstable Vietnam War veteran works ...          1976   
2  Intent on seeing the Cahulawassee River before...          1972   
3  King Arthur, accompanied by his squire, recrui...          1975   
4  12 American military prisoners in World War II...          1967   
5  A British sketch comedy series with the shows ...          1969   
6  Brian Cohen is an average young Jewish man, bu...          1979   
7  When a madman dubbed 'Scorpio' terrorizes San ...          1971   
8  In the 1930s, bored waitress Bonnie Parker fal...          1967   
9  Two small children and a ship's cook survive a...          1980   

  age_certification  runtime                                       genres  \
0             TV-MA       51                            ['documentation']   
1                 R      114                           ['drama', 'crime']   
2                 R      109  ['drama', 'action', 'thriller', 'european']   
3                PG       91              ['fantasy', 'action', 'comedy']   
4               NaN      150                            ['war', 'action']   
5             TV-14       30                       ['comedy', 'european']   
6                 R       94                                   ['comedy']   
7                 R      102              ['thriller', 'action', 'crime']   
8                 R      110                 ['crime', 'drama', 'action']   
9                 R      104               ['romance', 'action', 'drama']   

  production_countries  seasons    imdb_id  imdb_score  imdb_votes  \
0               ['US']      1.0        NaN         NaN         NaN   
1               ['US']      NaN  tt0075314         8.2    808582.0   
2               ['US']      NaN  tt0068473         7.7    107673.0   
3               ['GB']      NaN  tt0071853         8.2    534486.0   
4         ['GB', 'US']      NaN  tt0061578         7.7     72662.0   
5               ['GB']      4.0  tt0063929         8.8     73424.0   
6               ['GB']      NaN  tt0079470         8.0    395024.0   
7               ['US']      NaN  tt0066999         7.7    155051.0   
8               ['US']      NaN  tt0061418         7.7    112048.0   
9               ['US']      NaN  tt0080453         5.8     69844.0   

   tmdb_popularity  tmdb_score  
0            0.600         NaN  
1           40.965       8.179  
2           10.010       7.300  
3           15.461       7.811  
4           20.398       7.600  
5           17.617       8.306  
6           17.770       7.800  
7           12.817       7.500  
8           15.687       7.500  
9           50.324       6.156

(genres和production——countries单元格中含多个值)--应进行拆分

In [15]:
cleaned_titles["genres"][1]

"['drama', 'crime']"

(虽然genres表示形式是列表，但其实际类型并非字符串列表，而是字符串，无法直接用value_counts统计各个值出现的次数。
我们可以使用Python内置的eval函数，它可以把字符串转换成表达式，所以可以帮我们把表示列表的字符串转换成列表本身。)

In [16]:
cleaned_titles["genres"]=cleaned_titles["genres"].apply(lambda s:eval(s))

In [21]:
cleaned_titles["genres"][1]

['drama', 'crime']

(转换为列表后，就能用DataFrame的explode方法，把那个列的列表值拆分成单独的行。)

In [23]:
cleaned_titles=cleaned_titles.explode('genres')

In [24]:
cleaned_titles["genres"][1]

1    drama
1    crime
Name: genres, dtype: object

In [25]:
cleaned_titles.head(10)

id                                title   type  \
0  ts300399  Five Came Back: The Reference Films   SHOW   
1   tm84618                          Taxi Driver  MOVIE   
1   tm84618                          Taxi Driver  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   

                                         description  release_year  \
0  This collection includes 12 World War II-era p...          1945   
1  A mentally unstable Vietnam War veteran works ...          1976   
1  A mentally unstable Vietnam War veteran works ...          1976   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
3  King Arthur, accompanied by his squire, recrui...          1975   
3  King Arthur, accompanied by his squire, recrui...          1975   
3  King Arthur, accompanied by his squire, recrui...          1975   

  age_certification  runtime         genres production_countries  seasons  \
0             TV-MA       51  documentation               ['US']      1.0   
1                 R      114          drama               ['US']      NaN   
1                 R      114          crime               ['US']      NaN   
2                 R      109          drama               ['US']      NaN   
2                 R      109         action               ['US']      NaN   
2                 R      109       thriller               ['US']      NaN   
2                 R      109       european               ['US']      NaN   
3                PG       91        fantasy               ['GB']      NaN   
3                PG       91         action               ['GB']      NaN   
3                PG       91         comedy               ['GB']      NaN   

     imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0        NaN         NaN         NaN            0.600         NaN  
1  tt0075314         8.2    808582.0           40.965       8.179  
1  tt0075314         8.2    808582.0           40.965       8.179  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
3  tt0071853         8.2    534486.0           15.461       7.811  
3  tt0071853         8.2    534486.0           15.461       7.811  
3  tt0071853         8.2    534486.0           15.461       7.811

In [29]:
cleaned_titles['production_countries'][4]

4    ['GB', 'US']
4    ['GB', 'US']
Name: production_countries, dtype: object

In [30]:
cleaned_titles["production_countries"] = cleaned_titles["production_countries"].apply(lambda s:eval(s)) 

In [32]:
cleaned_titles["production_countries"][4]

4    [GB, US]
4    [GB, US]
Name: production_countries, dtype: object

In [33]:
cleaned_titles = cleaned_titles.explode("production_countries") 

In [34]:
cleaned_titles.head(10)

id                                title   type  \
0  ts300399  Five Came Back: The Reference Films   SHOW   
1   tm84618                          Taxi Driver  MOVIE   
1   tm84618                          Taxi Driver  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   

                                         description  release_year  \
0  This collection includes 12 World War II-era p...          1945   
1  A mentally unstable Vietnam War veteran works ...          1976   
1  A mentally unstable Vietnam War veteran works ...          1976   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
3  King Arthur, accompanied by his squire, recrui...          1975   
3  King Arthur, accompanied by his squire, recrui...          1975   
3  King Arthur, accompanied by his squire, recrui...          1975   

  age_certification  runtime         genres production_countries  seasons  \
0             TV-MA       51  documentation                   US      1.0   
1                 R      114          drama                   US      NaN   
1                 R      114          crime                   US      NaN   
2                 R      109          drama                   US      NaN   
2                 R      109         action                   US      NaN   
2                 R      109       thriller                   US      NaN   
2                 R      109       european                   US      NaN   
3                PG       91        fantasy                   GB      NaN   
3                PG       91         action                   GB      NaN   
3                PG       91         comedy                   GB      NaN   

     imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0        NaN         NaN         NaN            0.600         NaN  
1  tt0075314         8.2    808582.0           40.965       8.179  
1  tt0075314         8.2    808582.0           40.965       8.179  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
3  tt0071853         8.2    534486.0           15.461       7.811  
3  tt0071853         8.2    534486.0           15.461       7.811  
3  tt0071853         8.2    534486.0           15.461       7.811

(cleaned_titles 处理完成)

In [35]:
cleaned_credits.head(10)

person_id       id             name                      character   role
0       3748  tm84618   Robert De Niro                  Travis Bickle  ACTOR
1      14658  tm84618     Jodie Foster                  Iris Steensma  ACTOR
2       7064  tm84618    Albert Brooks                            Tom  ACTOR
3       3739  tm84618    Harvey Keitel        Matthew 'Sport' Higgins  ACTOR
4      48933  tm84618  Cybill Shepherd                          Betsy  ACTOR
5      32267  tm84618      Peter Boyle                         Wizard  ACTOR
6     519612  tm84618   Leonard Harris      Senator Charles Palantine  ACTOR
7      29068  tm84618   Diahnne Abbott                Concession Girl  ACTOR
8     519613  tm84618      Gino Ardito             Policeman at Rally  ACTOR
9       3308  tm84618  Martin Scorsese  Passenger Watching Silhouette  ACTOR

## 数据干净度 

In [36]:
cleaned_titles.info()

<class 'pandas.core.frame.DataFrame'>
Index: 17818 entries, 0 to 5849
Data columns (total 15 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id                    17818 non-null  object 
 1   title                 17817 non-null  object 
 2   type                  17818 non-null  object 
 3   description           17790 non-null  object 
 4   release_year          17818 non-null  int64  
 5   age_certification     10889 non-null  object 
 6   runtime               17818 non-null  int64  
 7   genres                17755 non-null  object 
 8   production_countries  17439 non-null  object 
 9   seasons               6224 non-null   float64
 10  imdb_id               17116 non-null  object 
 11  imdb_score            16976 non-null  float64
 12  imdb_votes            16945 non-null  float64
 13  tmdb_popularity       17663 non-null  float64
 14  tmdb_score            17241 non-null  float64
dtypes: float64(5), int64(2), 

从输出结果来看，cleaned_titles数据共有17818条观察值，
其中：title、description、age_certification、genres、production_countries、seasons、imdb_id、imdb_score、tmdb_popularity、tmdb_score、imdb_votes、tmdb_popularity、tmdb_score变量均存在缺失值，将在后续进行评估和清理。

（release——year表示年份，数据类型应为日期）---数字-->日期 转换

In [37]:
cleaned_titles["release_year"]=pd.to_datetime(cleaned_titles["release_year"],format="%Y")

In [38]:
cleaned_titles["release_year"]

0      1945-01-01
1      1976-01-01
1      1976-01-01
2      1972-01-01
2      1972-01-01
          ...    
5847   2021-01-01
5848   2021-01-01
5849   2021-01-01
5849   2021-01-01
5849   2021-01-01
Name: release_year, Length: 17818, dtype: datetime64[ns]

In [40]:
cleaned_credits.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 77801 entries, 0 to 77800
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   person_id  77801 non-null  int64 
 1   id         77801 non-null  object
 2   name       77801 non-null  object
 3   character  68029 non-null  object
 4   role       77801 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.0+ MB


(person_id 应为字符串 不应该是数字)（character存在缺失）

In [41]:
cleaned_credits["person_id"] = cleaned_credits["person_id"].astype("str")

In [42]:
cleaned_credits["person_id"]

0           3748
1          14658
2           7064
3           3739
4          48933
          ...   
77796     736339
77797     399499
77798     373198
77799     378132
77800    1950416
Name: person_id, Length: 77801, dtype: object

### 处理缺失数据

（在cleaned_titles中，title、description、age_certification、genres、production_countries、seasons、imdb_id、imdb_score、tmdb_popularity、tmdb_score、imdb_votes、tmdb_popularity、tmdb_score变量存在缺失值。


由于影视作品的标题、描述、适龄认证、发行国家、电视剧季数、IMDB的ID、TMDB的流行度、TMDB的评分，并不影响我们挖掘各个流派中的高IMDB评分作品演员，

可缺：title、description、age_certification、production_countries、seasons、imdb_id、tmdb_popularity、tmdb_score、imdb_votes、tmdb_popularity、tmdb_score变量值存在空缺的观察值。


但imdb_score和genres，即IMDB评分和流派，和我们后续要做的分析息息相关。

不可缺：imdb_score和genres

（提取出imdb_score缺失观察值进行查看。）

In [43]:
cleaned_titles.query("imdb_score.isnull()")

id                                      title   type  \
0      ts300399        Five Came Back: The Reference Films   SHOW   
75     tm132164                       Bill Hicks: Sane Man  MOVIE   
145    ts251477                            My First Errand   SHOW   
145    ts251477                            My First Errand   SHOW   
145    ts251477                            My First Errand   SHOW   
...         ...                                        ...    ...   
5810  tm1225897                                 Social Man  MOVIE   
5833   ts307884                                 HQ Barbers   SHOW   
5840  tm1216735                            Sun of the Soil  MOVIE   
5844  tm1074617              Bling Empire - The Afterparty  MOVIE   
5848  tm1035612  Dad Stop Embarrassing Me - The Afterparty  MOVIE   

                                            description release_year  \
0     This collection includes 12 World War II-era p...   1945-01-01   
75    Sane Man was filmed before Bill recorded ‘Dang...   1989-01-01   
145   “Hajimete no Otsukai” (First Errand) is a Japa...   1991-01-01   
145   “Hajimete no Otsukai” (First Errand) is a Japa...   1991-01-01   
145   “Hajimete no Otsukai” (First Errand) is a Japa...   1991-01-01   
...                                                 ...          ...   
5810  Two competitive social media Influencers go he...   2021-01-01   
5833  When a family run barber shop in the heart of ...   2021-01-01   
5840  In 14th-century Mali, an ambitious young royal...   2022-01-01   
5844  The stars of "Bling Empire" discuss the show's...   2021-01-01   
5848  Jamie Foxx, David Alan Grier and more from the...   2021-01-01   

     age_certification  runtime         genres production_countries  seasons  \
0                TV-MA       51  documentation                   US      1.0   
75                   R       80         comedy                   US      NaN   
145               TV-G       18  documentation                   JP     12.0   
145               TV-G       18         family                   JP     12.0   
145               TV-G       18        reality                   JP     12.0   
...                ...      ...            ...                  ...      ...   
5810               NaN       96          drama                  NaN      NaN   
5833             TV-14       24         comedy                   NG      1.0   
5840               NaN       26            NaN                  NaN      NaN   
5844               NaN       35            NaN                   US      NaN   
5848             PG-13       37            NaN                   US      NaN   

         imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0            NaN         NaN         NaN            0.600         NaN  
75           NaN         NaN         NaN            3.377         7.5  
145          NaN         NaN         NaN            7.730         7.8  
145          NaN         NaN         NaN            7.730         7.8  
145          NaN         NaN         NaN            7.730         7.8  
...          ...         ...         ...              ...         ...  
5810  tt20198164         NaN         NaN              NaN         NaN  
5833         NaN         NaN         NaN            0.840         NaN  
5840         NaN         NaN         NaN            1.179         7.0  
5844         NaN         NaN         NaN              NaN         NaN  
5848         NaN         NaN         NaN            1.296        10.0  

[842 rows x 15 columns]

由于缺失分析所需的核心数据imdb_score，(无法分析)我们将把这些观察值删除， 并查看删除后该列空缺值个数和：

In [44]:
cleaned_titles = cleaned_titles.dropna(subset=["imdb_score"])

In [46]:
cleaned_titles["imdb_score"].isnull().sum()

0

(提取genres缺失的观察值查看)

In [48]:
cleaned_titles.query("genres.isnull()")

id                                              title   type  \
1813    ts77824  My Next Guest Needs No Introduction With David...   SHOW   
1939   ts215037                              Minecraft: Story Mode   SHOW   
2386    ts74805                   A Little Help with Carol Burnett   SHOW   
2658   ts265844                                           #ABtalks   SHOW   
4274  tm1172010                                  The Lockdown Plan  MOVIE   
4648  tm1113921                                           In Vitro  MOVIE   

                                            description release_year  \
1813  TV legend David Letterman teams up with fascin...   2018-01-01   
1939  MInecraft: Story Mode is an interactive, anima...   2018-01-01   
2386  In this unscripted series starring comedy lege...   2018-01-01   
2658  #ABtalks is a YouTube interview show hosted by...   2018-01-01   
4274                                                NaN   2020-01-01   
4648  'In Vitro' is an otherworldly rumination on me...   2019-01-01   

     age_certification  runtime genres production_countries  seasons  \
1813             TV-MA       50    NaN                   US      4.0   
1939             TV-PG       52    NaN                   US      1.0   
2386              TV-G       24    NaN                   US      1.0   
2658             TV-PG       68    NaN                  NaN      1.0   
4274               NaN       49    NaN                  NaN      NaN   
4648               NaN       27    NaN                  NaN      NaN   

         imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
1813   tt7829834         7.8      5581.0            8.217         7.6  
1939  tt10498322         5.6       347.0              NaN         NaN  
2386   tt7204366         6.3       237.0            1.621         6.2  
2658  tt12635254         9.6         7.0              NaN         NaN  
4274  tt13079112         6.5         NaN              NaN         NaN  
4648  tt10545994         7.7         NaN              NaN         NaN

(缺乏核心数据 无法分析)

In [50]:
cleaned_titles = cleaned_titles.dropna(subset=["genres"])

In [51]:
cleaned_titles["genres"].isnull().sum()

0

In [52]:
cleaned_credits.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 77801 entries, 0 to 77800
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   person_id  77801 non-null  object
 1   id         77801 non-null  object
 2   name       77801 non-null  object
 3   character  68029 non-null  object
 4   role       77801 non-null  object
dtypes: object(5)
memory usage: 3.0+ MB


(cleaned_credits中只有 character存在缺失值) 角色名并不影响对演员的分析（可忽略）

### 处理重复数据

(根据数据变量的含义以及内容来看，cleaned_titles里不应该存在每个变量值都相同的观察值，因此查看是否存在重复值。)

In [53]:
cleaned_titles.duplicated().sum()

0

### 处理不一致数据

（针对cleaned_titles，不一致数据可能存在于genres和character变量中，我们将查看是否存在多个不同值指代同一流派，以及多个不同值指代同一国家的情况。）

In [54]:
cleaned_titles["genres"].value_counts()

genres
drama            3357
comedy           2419
thriller         1446
action           1339
romance          1080
crime            1066
documentation     981
family            769
animation         732
fantasy           727
european          679
scifi             647
horror            438
history           336
music             266
reality           226
war               221
sport             188
western            53
Name: count, dtype: int64

In [56]:
cleaned_titles = cleaned_titles.query('genres!=""')
len(cleaned_titles.query('genres ==""'))

0

In [58]:
cleaned_titles['production_countries'].value_counts()

production_countries
US    5648
IN    1610
GB    1068
JP    1046
FR     720
      ... 
GT       1
CU       1
LK       1
NP       1
FO       1
Name: count, Length: 108, dtype: int64

In [61]:
with pd.option_context("display.max_rows",None):
    print(cleaned_titles["production_countries"].value_counts())

production_countries
US         5648
IN         1610
GB         1068
JP         1046
FR          720
KR          637
ES          637
CA          608
DE          383
CN          295
MX          264
IT          224
BR          221
AU          217
TR          195
PH          192
AR          150
ID          149
BE          148
TW          133
NG          131
PL          126
ZA          103
NL          102
HK          102
CO           94
EG           93
DK           89
TH           87
SE           81
LB           70
NO           68
AE           52
IE           49
SG           47
XX           43
IL           42
RU           41
CL           35
CH           33
PS           32
BG           31
MY           30
SA           28
AT           28
IS           28
LU           27
NZ           27
PE           26
RO           25
QA           24
CZ           22
JO           19
FI           18
HU           18
UY           15
MA           15
PT           14
KH           10
KW           10
PR            9
PK 

(LB和Lebanon都在表示同一国家，需要进行统一。)

In [63]:
cleaned_titles["production_countries"] = cleaned_titles["production_countries"].replace({"Lebanon":"LB"})

In [65]:
with pd.option_context("display.max_rows",None):
    print(cleaned_titles.explode("production_countries")["production_countries"].value_counts())

production_countries
US    5648
IN    1610
GB    1068
JP    1046
FR     720
KR     637
ES     637
CA     608
DE     383
CN     295
MX     264
IT     224
BR     221
AU     217
TR     195
PH     192
AR     150
ID     149
BE     148
TW     133
NG     131
PL     126
ZA     103
HK     102
NL     102
CO      94
EG      93
DK      89
TH      87
SE      81
LB      71
NO      68
AE      52
IE      49
SG      47
XX      43
IL      42
RU      41
CL      35
CH      33
PS      32
BG      31
MY      30
AT      28
SA      28
IS      28
LU      27
NZ      27
PE      26
RO      25
QA      24
CZ      22
JO      19
FI      18
HU      18
UY      15
MA      15
PT      14
KW      10
KH      10
PK       9
PR       9
UA       8
MT       8
VN       8
LT       7
IR       7
CD       7
SU       7
TN       7
SN       6
AL       6
KE       6
GH       6
IQ       5
MU       5
CY       5
KN       4
GR       4
IO       4
SY       4
TZ       4
MC       4
GL       3
CM       3
AO       3
BS       3
HR       3
BD       3


(针对cleaned_credits，不一致数据可能存在于role中，我们将查看是否存在多个不同值指代同一演职员类型的情况。)

In [66]:
cleaned_credits["role"].value_counts()

role
ACTOR       73251
DIRECTOR     4550
Name: count, dtype: int64

role只有两种可能的值，ACTOR或DIRECTOR，不存在不一致数据。我们可以把这列的类型转换为Category，好处是比字符串类型更节约内存空间，也能表明说值的类型有限。

In [67]:
cleaned_credits["role"] = cleaned_credits["role"].astype("category")

In [68]:
cleaned_credits["role"]

0           ACTOR
1           ACTOR
2           ACTOR
3           ACTOR
4           ACTOR
           ...   
77796       ACTOR
77797       ACTOR
77798       ACTOR
77799       ACTOR
77800    DIRECTOR
Name: role, Length: 77801, dtype: category
Categories (2, object): ['ACTOR', 'DIRECTOR']

### 处理无效或错误数据

In [69]:
cleaned_titles.describe()

release_year       runtime      seasons    imdb_score  \
count                          16970  16970.000000  5954.000000  16970.000000   
mean   2015-11-14 22:42:51.974072064     80.912552     2.455492      6.514207   
min              1954-01-01 00:00:00      0.000000     1.000000      1.500000   
25%              2015-01-01 00:00:00     45.000000     1.000000      5.800000   
50%              2018-01-01 00:00:00     90.000000     2.000000      6.600000   
75%              2020-01-01 00:00:00    107.000000     3.000000      7.300000   
max              2022-01-01 00:00:00    225.000000    42.000000      9.500000   
std                              NaN     39.596172     2.869428      1.131095   

         imdb_votes  tmdb_popularity    tmdb_score  
count  1.694100e+04     16842.000000  16515.000000  
mean   3.281655e+04        29.396307      6.846933  
min    5.000000e+00         0.600000      1.000000  
25%    7.800000e+02         4.070000      6.200000  
50%    3.508000e+03        10.195000      6.900000  
75%    1.697800e+04        23.639000      7.500000  
max    2.294231e+06      2274.044000     10.000000  
std    1.141492e+05        93.178235      1.078831

(从以上统计信息来看，original_titles里不存在脱离现实意义的数值)

# 整理数据

（以ID作为连接的键）（采用内连接方式 ）

In [71]:
credits_with_titles = pd.merge(cleaned_credits,cleaned_titles,on="id",how="inner")

In [72]:
credits_with_titles.head(10)

person_id       id             name                character   role  \
0      3748  tm84618   Robert De Niro            Travis Bickle  ACTOR   
1      3748  tm84618   Robert De Niro            Travis Bickle  ACTOR   
2     14658  tm84618     Jodie Foster            Iris Steensma  ACTOR   
3     14658  tm84618     Jodie Foster            Iris Steensma  ACTOR   
4      7064  tm84618    Albert Brooks                      Tom  ACTOR   
5      7064  tm84618    Albert Brooks                      Tom  ACTOR   
6      3739  tm84618    Harvey Keitel  Matthew 'Sport' Higgins  ACTOR   
7      3739  tm84618    Harvey Keitel  Matthew 'Sport' Higgins  ACTOR   
8     48933  tm84618  Cybill Shepherd                    Betsy  ACTOR   
9     48933  tm84618  Cybill Shepherd                    Betsy  ACTOR   

         title   type                                        description  \
0  Taxi Driver  MOVIE  A mentally unstable Vietnam War veteran works ...   
1  Taxi Driver  MOVIE  A mentally unstable Vietnam War veteran works ...   
2  Taxi Driver  MOVIE  A mentally unstable Vietnam War veteran works ...   
3  Taxi Driver  MOVIE  A mentally unstable Vietnam War veteran works ...   
4  Taxi Driver  MOVIE  A mentally unstable Vietnam War veteran works ...   
5  Taxi Driver  MOVIE  A mentally unstable Vietnam War veteran works ...   
6  Taxi Driver  MOVIE  A mentally unstable Vietnam War veteran works ...   
7  Taxi Driver  MOVIE  A mentally unstable Vietnam War veteran works ...   
8  Taxi Driver  MOVIE  A mentally unstable Vietnam War veteran works ...   
9  Taxi Driver  MOVIE  A mentally unstable Vietnam War veteran works ...   

  release_year age_certification  runtime genres production_countries  \
0   1976-01-01                 R      114  drama                   US   
1   1976-01-01                 R      114  crime                   US   
2   1976-01-01                 R      114  drama                   US   
3   1976-01-01                 R      114  crime                   US   
4   1976-01-01                 R      114  drama                   US   
5   1976-01-01                 R      114  crime                   US   
6   1976-01-01                 R      114  drama                   US   
7   1976-01-01                 R      114  crime                   US   
8   1976-01-01                 R      114  drama                   US   
9   1976-01-01                 R      114  crime                   US   

   seasons    imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0      NaN  tt0075314         8.2    808582.0           40.965       8.179  
1      NaN  tt0075314         8.2    808582.0           40.965       8.179  
2      NaN  tt0075314         8.2    808582.0           40.965       8.179  
3      NaN  tt0075314         8.2    808582.0           40.965       8.179  
4      NaN  tt0075314         8.2    808582.0           40.965       8.179  
5      NaN  tt0075314         8.2    808582.0           40.965       8.179  
6      NaN  tt0075314         8.2    808582.0           40.965       8.179  
7      NaN  tt0075314         8.2    808582.0           40.965       8.179  
8      NaN  tt0075314         8.2    808582.0           40.965       8.179  
9      NaN  tt0075314         8.2    808582.0           40.965       8.179

In [73]:
actor_with_titles = credits_with_titles.query("role=='ACTOR'")

In [76]:
groupby_genres_and_person_id = actor_with_titles.groupby(["genres","person_id"])

In [77]:
imdb_score_groupby_genres_and_person_id= groupby_genres_and_person_id["imdb_score"].mean()

In [78]:
imdb_score_groupby_genres_and_person_id

genres   person_id
action   1000         6.866667
         100007       7.000000
         100013       6.400000
         100019       6.500000
         100020       6.500000
                        ...   
western  993735       6.500000
         998673       7.300000
         998674       7.300000
         998675       7.300000
         99940        4.000000
Name: imdb_score, Length: 168881, dtype: float64

In [79]:
imdb_score_groupby_genres_and_person_id_df= imdb_score_groupby_genres_and_person_id.reset_index()

In [80]:
imdb_score_groupby_genres_and_person_id_df

genres person_id  imdb_score
0        action      1000    6.866667
1        action    100007    7.000000
2        action    100013    6.400000
3        action    100019    6.500000
4        action    100020    6.500000
...         ...       ...         ...
168876  western    993735    6.500000
168877  western    998673    7.300000
168878  western    998674    7.300000
168879  western    998675    7.300000
168880  western     99940    4.000000

[168881 rows x 3 columns]

In [83]:
genres_max_score = imdb_score_groupby_genres_and_person_id_df.groupby("genres")["imdb_score"].max()
genres_max_score 

genres
action           9.3
animation        9.3
comedy           9.2
crime            9.5
documentation    9.1
drama            9.5
european         8.9
family           9.3
fantasy          9.3
history          9.1
horror           9.0
music            8.8
reality          8.9
romance          9.2
scifi            9.3
sport            9.1
thriller         9.5
war              8.8
western          8.9
Name: imdb_score, dtype: float64

In [84]:
genres_max_score_with_person_id = pd.merge(imdb_score_groupby_genres_and_person_id_df,genres_max_score,on=["genres","imdb_score"])
genres_max_score_with_person_id

genres person_id  imdb_score
0     action     12790         9.3
1     action      1303         9.3
2     action     21033         9.3
3     action    336830         9.3
4     action     86591         9.3
..       ...       ...         ...
131      war    826547         8.8
132  western     22311         8.9
133  western     28166         8.9
134  western     28180         8.9
135  western     93017         8.9

[136 rows x 3 columns]

In [86]:
actor_id_with_names=cleaned_credits[["person_id","name"]].drop_duplicates()

In [87]:
actor_id_with_names.head(10)

person_id             name
0      3748   Robert De Niro
1     14658     Jodie Foster
2      7064    Albert Brooks
3      3739    Harvey Keitel
4     48933  Cybill Shepherd
5     32267      Peter Boyle
6    519612   Leonard Harris
7     29068   Diahnne Abbott
8    519613      Gino Ardito
9      3308  Martin Scorsese

In [88]:
genres_max_score_with_actor_name=pd.merge(genres_max_score_with_person_id,actor_id_with_names,on="person_id")

In [89]:
genres_max_score_with_actor_name

genres person_id  imdb_score                name
0     action     12790         9.3         Olivia Hack
1     action      1303         9.3       Jessie Flower
2     action     21033         9.3          Zach Tyler
3     action    336830         9.3     André Sogliuzzo
4     action     86591         9.3       Cricket Leigh
..       ...       ...         ...                 ...
131      war    826547         8.8         Yuto Uemura
132  western     22311         8.9     Koichi Yamadera
133  western     28166         8.9  Megumi Hayashibara
134  western     28180         8.9      Unsho Ishizuka
135  western     93017         8.9            Aoi Tada

[136 rows x 4 columns]

In [93]:
genres_max_score_with_actor_name = genres_max_score_with_actor_name.sort_values("genres").reset_index().drop("index",axis=1)

In [94]:
genres_max_score_with_actor_name 

genres person_id  imdb_score                name
0     action     12790         9.3         Olivia Hack
1     action      1303         9.3       Jessie Flower
2     action     21033         9.3          Zach Tyler
3     action    336830         9.3     André Sogliuzzo
4     action     86591         9.3       Cricket Leigh
..       ...       ...         ...                 ...
131      war    826547         8.8         Yuto Uemura
132  western     28180         8.9      Unsho Ishizuka
133  western     22311         8.9     Koichi Yamadera
134  western     28166         8.9  Megumi Hayashibara
135  western     93017         8.9            Aoi Tada

[136 rows x 4 columns]